environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentModificationService.java

```diff
@@ -0,0 +1,162 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+import org.springframework.util.CollectionUtils;
+
+import com.sequenceiq.cloudbreak.cloud.model.CloudRegions;
+import com.sequenceiq.cloudbreak.common.exception.NotFoundException;
+import com.sequenceiq.cloudbreak.util.ValidationResult;
+import com.sequenceiq.environment.CloudPlatform;
+import com.sequenceiq.environment.api.v1.environment.model.request.EnvironmentChangeCredentialRequest;
+import com.sequenceiq.environment.credential.domain.Credential;
+import com.sequenceiq.environment.credential.service.CredentialService;
+import com.sequenceiq.environment.environment.domain.Environment;
+import com.sequenceiq.environment.environment.domain.Region;
+import com.sequenceiq.environment.environment.dto.EnvironmentDto;
+import com.sequenceiq.environment.environment.dto.EnvironmentDtoConverter;
+import com.sequenceiq.environment.environment.dto.EnvironmentEditDto;
+import com.sequenceiq.environment.environment.dto.LocationDto;
+import com.sequenceiq.environment.environment.repository.EnvironmentRepository;
+import com.sequenceiq.environment.network.NetworkService;
+import com.sequenceiq.environment.network.domain.BaseNetwork;
+import com.sequenceiq.environment.network.dto.NetworkDto;
+
+@Service
+public class EnvironmentModificationService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(EnvironmentModificationService.class);
+
+    private final EnvironmentDtoConverter environmentDtoConverter;
+
+    private final EnvironmentRepository environmentRepository;
+
+    private final EnvironmentService environmentService;
+
+    private final CredentialService credentialService;
+
+    private final NetworkService networkService;
+
+    public EnvironmentModificationService(EnvironmentDtoConverter environmentDtoConverter, EnvironmentRepository environmentRepository,
+            EnvironmentService environmentService, CredentialService credentialService, NetworkService networkService) {
+        this.environmentDtoConverter = environmentDtoConverter;
+        this.environmentRepository = environmentRepository;
+        this.environmentService = environmentService;
+        this.credentialService = credentialService;
+        this.networkService = networkService;
+    }
+
+    public EnvironmentDto edit(String environmentName, EnvironmentEditDto editDto) {
+        Environment env = environmentService.getByNameForAccountId(environmentName, editDto.getAccountId());
+        editDescriptionIfChanged(env, editDto);
+        editLocationAndRegionsIfChanged(env, editDto);
+        editNetworkIfChanged(env, editDto);
+        Environment saved = environmentRepository.save(env);
+        return environmentDtoConverter.environmentToDto(saved);
+    }
+
+    public EnvironmentDto changeCredential(String accountId, String environmentName, EnvironmentChangeCredentialRequest credentialChangeRequest) {
```

Comment:
Same principle applies here. Services must not know anything about the Controller layer and their request objects. It's the controller's responsibility to convert between dtos and requests/responses with some converter. This separation of concerns makes it possible to use these services elsewhere. (e.g. in other services or other api version controllers.)

Reply:
already done :D

Reply:
`EnvironmentChangeCredentialRequest`